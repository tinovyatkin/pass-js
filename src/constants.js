/**
 * Common constants for fields names and values:
 * https://developer.apple.com/library/content/documentation/UserExperience/Reference/PassKit_Bundle/Chapters/LowerLevel.html#//apple_ref/doc/uid/TP40012026-CH3-SW3
 *
 */

'use strict';

const PASS_MIME_TYPE = 'application/vnd.apple.pkpass';
module.exports.PASS_MIME_TYPE = PASS_MIME_TYPE;

const TRANSIT = {
  AIR: 'PKTransitTypeAir',
  BOAT: 'PKTransitTypeBoat',
  BUS: 'PKTransitTypeBus',
  TRAIN: 'PKTransitTypeTrain',
  GENERIC: 'PKTransitTypeGeneric',
};
module.exports.TRANSIT = TRANSIT;

const textDirection = {
  LEFT: 'PKTextAlignmentLeft',
  CENTER: 'PKTextAlignmentCenter',
  RIGHT: 'PKTextAlignmentRight',
  NATURAL: 'PKTextAlignmentNatural',
};
module.exports.textDirection = textDirection;

const barcodeFormat = {
  QR: 'PKBarcodeFormatQR',
  PDF417: 'PKBarcodeFormatPDF417',
  Aztec: 'PKBarcodeFormatAztec',
  Code128: 'PKBarcodeFormatCode128',
};
module.exports.barcodeFormat = barcodeFormat;

const dateTimeFormat = {
  NONE: 'PKDateStyleNone',
  SHORT: 'PKDateStyleShort',
  MEDIUM: 'PKDateStyleMedium',
  LONG: 'PKDateStyleLong',
  FULL: 'PKDateStyleFull',
};
module.exports.dateTimeFormat = dateTimeFormat;

const dataDetector = {
  PHONE: 'PKDataDetectorTypePhoneNumber',
  LINK: 'PKDataDetectorTypeLink',
  ADDRESS: 'PKDataDetectorTypeAddress',
  CALENDAR: 'PKDataDetectorTypeCalendarEvent',
};
module.exports.dataDetector = dataDetector;

// Supported images.
const IMAGES = {
  icon: {
    width: 29,
    height: 29,
    required: true,
  },
  logo: {
    width: 160,
    height: 50,
    required: true,
  },
  background: {
    width: 180,
    height: 220,
  },
  footer: {
    width: 295,
    height: 15,
  },
  strip: {
    width: 375,
    height: 123,
  },
  thumbnail: {
    width: 90,
    height: 90,
  },
};

const DENSITIES = ['1x', '2x', '3x'];
module.exports.IMAGES = IMAGES;
module.exports.DENSITIES = DENSITIES;

// Supported passbook styles.
const PASS_STYLES = [
  'boardingPass',
  'coupon',
  'eventTicket',
  'storeCard',
  'generic',
];
module.exports.PASS_STYLES = PASS_STYLES;

// Optional top level fields
// Top-level pass fields.
// https://developer.apple.com/library/content/documentation/UserExperience/Reference/PassKit_Bundle/Chapters/TopLevel.html#//apple_ref/doc/uid/TP40012026-CH2-SW1
const TOP_LEVEL_FIELDS = {
  // Standard Keys
  description: {
    required: true,
    type: 'string',
    templatable: true,
    localizable: true,
  },
  organizationName: {
    required: true,
    type: 'string',
    templatable: true,
    localizable: true,
  },
  passTypeIdentifier: {
    required: true,
    type: 'string',
    templatable: true,
  },
  serialNumber: {
    required: true,
    type: 'string',
  },
  teamIdentifier: {
    required: true,
    type: 'string',
    templatable: true,
  },
  associatedStoreIdentifiers: {
    required: false,
    type: Array,
    templatable: true,
  },
  // Expiration Keys
  expirationDate: {
    type: 'string', // W3C date, as a string
  },
  voided: {
    type: Boolean,
  },
  // Relevance Keys
  beacons: {
    type: Array,
  },
  locations: {
    type: Array,
  },
  maxDistance: {
    type: 'number',
  },
  relevantDate: {
    type: 'string', // W3C date, as a string
  },
  // Visual Appearance Keys
  barcodes: {
    type: Array,
  },
  backgroundColor: {
    type: 'string',
    templatable: true,
  },
  foregroundColor: {
    type: 'string',
    templatable: true,
  },
  groupingIdentifier: {
    type: 'string',
  },
  labelColor: {
    type: 'string',
    templatable: true,
  },
  logoText: {
    type: 'string',
    templatable: true,
    localizable: true,
  },
  suppressStripShine: {
    type: Boolean,
    templatable: true,
  },
  // Web Service Keys
  authenticationToken: {
    type: 'string',
    minlength: 16,
  },
  webServiceURL: {
    type: 'string',
    templatable: true,
  },
  // NFC-Enabled Pass Keys
  nfc: {
    type: Object,
  },
};
module.exports.TOP_LEVEL_FIELDS = TOP_LEVEL_FIELDS;

// Pass structure keys.
// https://developer.apple.com/library/content/documentation/UserExperience/Reference/PassKit_Bundle/Chapters/LowerLevel.html#//apple_ref/doc/uid/TP40012026-CH3-SW3
const STRUCTURE_FIELDS = [
  'auxiliaryFields',
  'backFields',
  'headerFields',
  'primaryFields',
  'secondaryFields',
  'transitType',
];
module.exports.STRUCTURE_FIELDS = STRUCTURE_FIELDS;

// These images are required for a valid pass.
const REQUIRED_IMAGES = ['icon', 'logo'];
module.exports.REQUIRED_IMAGES = REQUIRED_IMAGES;
