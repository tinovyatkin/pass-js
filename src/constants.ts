/**
 * Common constants for fields names and values:
 * https://developer.apple.com/library/content/documentation/UserExperience/Reference/PassKit_Bundle/Chapters/LowerLevel.html#//apple_ref/doc/uid/TP40012026-CH3-SW3
 *
 */

'use strict';

import { PassStyle, TransitType } from './interfaces';

export const PASS_MIME_TYPE = 'application/vnd.apple.pkpass';

export const TRANSIT = {
  AIR: 'PKTransitTypeAir' as TransitType,
  BOAT: 'PKTransitTypeBoat' as TransitType,
  BUS: 'PKTransitTypeBus' as TransitType,
  TRAIN: 'PKTransitTypeTrain' as TransitType,
  GENERIC: 'PKTransitTypeGeneric' as TransitType,
};

export const textDirection = {
  LEFT: 'PKTextAlignmentLeft',
  CENTER: 'PKTextAlignmentCenter',
  RIGHT: 'PKTextAlignmentRight',
  NATURAL: 'PKTextAlignmentNatural',
};

export const barcodeFormat = {
  QR: 'PKBarcodeFormatQR',
  PDF417: 'PKBarcodeFormatPDF417',
  Aztec: 'PKBarcodeFormatAztec',
  Code128: 'PKBarcodeFormatCode128',
};

export const dateTimeFormat = {
  NONE: 'PKDateStyleNone',
  SHORT: 'PKDateStyleShort',
  MEDIUM: 'PKDateStyleMedium',
  LONG: 'PKDateStyleLong',
  FULL: 'PKDateStyleFull',
};

export const dataDetector = {
  PHONE: 'PKDataDetectorTypePhoneNumber',
  LINK: 'PKDataDetectorTypeLink',
  ADDRESS: 'PKDataDetectorTypeAddress',
  CALENDAR: 'PKDataDetectorTypeCalendarEvent',
};

// Supported images.
/** @type {{[k: string]: { width: number, height: number, required?: boolean }}} */
export const IMAGES = {
  icon: {
    width: 29,
    height: 29,
    required: true,
  },
  logo: {
    width: 160,
    height: 50,
    required: true,
  },
  background: {
    width: 180,
    height: 220,
  },
  footer: {
    width: 295,
    height: 15,
  },
  strip: {
    width: 375,
    height: 123,
  },
  thumbnail: {
    width: 90,
    height: 90,
  },
};

export const DENSITIES = new Set(['1x', '2x', '3x']);

// Supported passbook styles.
export const PASS_STYLES = new Set([
  'boardingPass',
  'coupon',
  'eventTicket',
  'storeCard',
  'generic',
]) as Set<PassStyle>;

// Optional top level fields
// Top-level pass fields.
// https://developer.apple.com/library/content/documentation/UserExperience/Reference/PassKit_Bundle/Chapters/TopLevel.html#//apple_ref/doc/uid/TP40012026-CH2-SW1
export const TOP_LEVEL_FIELDS = {
  // Standard Keys
  description: {
    required: true,
    type: 'string',
    templatable: true,
    localizable: true,
  },
  organizationName: {
    required: true,
    type: 'string',
    templatable: true,
    localizable: true,
  },
  passTypeIdentifier: {
    required: true,
    type: 'string',
    templatable: true,
  },
  serialNumber: {
    required: true,
    type: 'string',
  },
  teamIdentifier: {
    required: true,
    type: 'string',
    templatable: true,
  },
  associatedStoreIdentifiers: {
    required: false,
    type: Array,
    templatable: true,
  },
  // Expiration Keys
  expirationDate: {
    type: 'string', // W3C date, as a string
  },
  voided: {
    type: Boolean,
  },
  // Relevance Keys
  beacons: {
    type: Array,
  },
  locations: {
    type: Array,
  },
  maxDistance: {
    type: 'number',
  },
  relevantDate: {
    type: 'string', // W3C date, as a string
  },
  // Visual Appearance Keys
  barcodes: {
    type: Array,
  },
  backgroundColor: {
    type: 'string',
    templatable: true,
  },
  foregroundColor: {
    type: 'string',
    templatable: true,
  },
  groupingIdentifier: {
    type: 'string',
  },
  labelColor: {
    type: 'string',
    templatable: true,
  },
  logoText: {
    type: 'string',
    templatable: true,
    localizable: true,
  },
  suppressStripShine: {
    type: Boolean,
    templatable: true,
  },
  // Web Service Keys
  authenticationToken: {
    type: 'string',
    minlength: 16,
  },
  webServiceURL: {
    type: 'string',
    templatable: true,
  },
  // NFC-Enabled Pass Keys
  nfc: {
    type: Object,
  },
};

// Pass structure keys.
// https://developer.apple.com/library/content/documentation/UserExperience/Reference/PassKit_Bundle/Chapters/LowerLevel.html#//apple_ref/doc/uid/TP40012026-CH3-SW3
export const STRUCTURE_FIELDS = [
  'auxiliaryFields',
  'backFields',
  'headerFields',
  'primaryFields',
  'secondaryFields',
] as readonly string[];

/** @type {Set.<'PKBarcodeFormatQR' | 'PKBarcodeFormatPDF417' | 'PKBarcodeFormatAztec' | 'PKBarcodeFormatCode128'>} */
export const BARCODES_FORMAT = new Set([
  'PKBarcodeFormatQR',
  'PKBarcodeFormatPDF417',
  'PKBarcodeFormatAztec',
  'PKBarcodeFormatCode128',
]);
