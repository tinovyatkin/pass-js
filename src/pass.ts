// Generate a pass file.

'use strict';

import { toBuffer as createZip } from 'do-not-zip';

import { getBufferHash } from './lib/getBufferHash';
import { PassImages } from './lib/images';
import { signManifest } from './lib/signManifest-forge';
import { PassBase } from './lib/base-pass';
import { ApplePass } from './interfaces';

/**
 * @typedef {'PKBarcodeFormatQR' | 'PKBarcodeFormatPDF417' | 'PKBarcodeFormatAztec' | 'PKBarcodeFormatCode128'} BarcodeFormat
 * @typedef {{format: BarcodeFormat, message: string, messageEncoding: string}} BarcodeDescriptor
 */

// Create a new pass.
//
// template  - The template
// fields    - Pass fields (description, serialNumber, logoText)
export class Pass extends PassBase {
  private template: import('./template').Template;
  /**
   *
   * @param {import('./template')} template
   * @param {*} [fields]
   * @param {*} [images]
   */
  constructor(
    template: import('./template').Template,
    fields: Partial<ApplePass> = {},
    images?: PassImages,
  ) {
    super(fields, images);

    this.template = template;
    if (images) Object.assign(this.images, images);

    Object.preventExtensions(this);
  }

  // Validate pass, throws error if missing a mandatory top-level field or image.
  validate(): void {
    // Check required top level fields
    for (const requiredField of [
      'description',
      'organizationName',
      'passTypeIdentifier',
      'serialNumber',
      'teamIdentifier',
    ])
      if (!(requiredField in this.fields))
        throw new ReferenceError(`${requiredField} is required in a Pass`);

    // authenticationToken && webServiceURL must be either both or none
    if ('webServiceURL' in this.fields) {
      if (typeof this.fields.authenticationToken !== 'string')
        throw new Error(
          'While webServiceURL is present, authenticationToken also required!',
        );
      if (this.fields.authenticationToken.length < 16)
        throw new ReferenceError(
          'authenticationToken must be at least 16 characters long!',
        );
    } else if ('authenticationToken' in this.fields)
      throw new TypeError(
        'authenticationToken is presented in Pass data while webServiceURL is missing!',
      );

    this.images.validate();
  }

  /**
   * Returns Pass as a Buffer
   *
   * @memberof Pass
   * @returns {Promise.<Buffer>}
   */
  async asBuffer(): Promise<Buffer> {
    // Validate before attempting to create
    this.validate();
    if (!this.template.certificate)
      throw new ReferenceError(
        `Set pass certificate in template before producing pass buffers`,
      );
    if (!this.template.key)
      throw new ReferenceError(
        `Set private key in pass template before producing pass buffers`,
      );

    // Creating new Zip file
    const zip = [] as { path: string; data: Buffer | string }[];

    // Adding required files
    // Create pass.json
    zip.push({ path: 'pass.json', data: JSON.stringify(this) });

    // Localization
    zip.push(...this.localization.toArray());

    // Images
    zip.push(...(await this.images.toArray()));

    // adding manifest
    // Construct manifest here
    const manifestJson = JSON.stringify(
      zip.reduce(
        (res, { path, data }) => {
          res[path] = getBufferHash(data);
          return res;
        },
        {} as { [k: string]: string },
      ),
    );
    zip.push({ path: 'manifest.json', data: manifestJson });

    // Create signature
    const signature = signManifest(
      this.template.certificate,
      this.template.key,
      manifestJson,
    );
    zip.push({ path: 'signature', data: signature });

    // finished!
    return createZip(zip);
  }
}
