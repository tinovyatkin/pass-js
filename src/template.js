/**
 * Passbook are created from templates
 */

'use strict';

const colorString = require('color-string');
const decodePrivateKey = require('./lib/decodePrivateKey');
const http2 = require('http2');
const path = require('path');
const { join } = require('path');
const { stat, readFile } = require('fs').promises;

const Pass = require('./pass');
const PassImages = require('./lib/images');
const { PASS_STYLES } = require('./constants');

const { HTTP2_HEADER_METHOD, HTTP2_HEADER_PATH } = http2.constants;

// Create a new template.
//
// style  - Pass style (coupon, eventTicket, etc)
// fields - Pass fields (passTypeIdentifier, teamIdentifier, etc)
class Template {
  constructor(style, fields = {}) {
    if (!PASS_STYLES.includes(style))
      throw new Error(`Unsupported pass style ${style}`);

    this.style = style;
    this.fields = {};
    // we will set all fields via class setters, as in the future we will implement strict validators
    // values validation: https://developer.apple.com/library/content/documentation/UserExperience/Reference/PassKit_Bundle/Chapters/TopLevel.html
    for (const [field, value] of Object.entries(fields)) {
      if (typeof this[field] === 'function') this[field](value);
    }

    if (style in fields) {
      Object.assign(this.fields, { [style]: fields[style] });
    }

    this.keysPath = 'keys';
    this.password = null;
    this.apn = null;
    this.images = new PassImages();
    Object.preventExtensions(this);
  }

  async pushUpdates(pushToken) {
    // https://developer.apple.com/library/content/documentation/UserExperience/Conceptual/PassKit_PG/Updating.html
    if (!this.apn || this.apn.destroyed) {
      // prepare certificate
      // creating APN Provider
      const identifier = this.passTypeIdentifier().replace(/^pass./, '');

      const cert = await readFile(
        path.resolve(this.keysPath, `${identifier}.pem`),
        'utf8',
      );

      /** @type {string} */
      const key = decodePrivateKey(cert, this.password, true);

      this.apn = http2.connect('https://api.push.apple.com:443', {
        key,
        cert,
      });
      // Events
      this.apn.once('goaway', () => this.apn.destroy());
      await new Promise((resolve, reject) => {
        this.apn.once('connect', resolve);
        this.apn.once('error', reject);
      });
    }

    // sending to APN
    return new Promise((resolve, reject) => {
      const req = this.apn.request({
        [HTTP2_HEADER_METHOD]: 'POST',
        [HTTP2_HEADER_PATH]: `/3/device/${encodeURIComponent(pushToken)}`,
      });

      // Cancel request after timeout
      req.setTimeout(5000, () => req.rstWithCancel());

      // Response handling
      req.on('response', headers => {
        // consuming data, even if we are not interesting in it
        req.on('data', () => {});
        req.on('end', () => resolve(headers));
      });

      // Error handling
      req.on('error', reject);
      req.on('timeout', () =>
        reject(new Error(`http2: timeout connecting to api.push.apple.com`)),
      );

      // Post payload (always empty in our case)
      req.end('{}');
    });
  }

  /**
   * Validates if given string is a correct color value for Pass fields
   *
   * @static
   * @param {string} value - a CSS color value, like 'red', '#fff', etc
   * @throws - if value is invalid this function will throw
   * @returns {string} - value converted to "rgb(222, 33, 22)" string
   * @memberof Template
   */
  static convertToRgb(value) {
    const rgb = colorString.get.rgb(value);
    if (rgb === null) throw new Error(`Invalid color value ${value}`);
    // convert to rgb(), stripping alpha channel
    return colorString.to.rgb(rgb.slice(0, 3));
  }

  passTypeIdentifier(v) {
    if (arguments.length === 1) {
      this.fields.passTypeIdentifier = v;
      return this;
    }
    return this.fields.passTypeIdentifier;
  }

  teamIdentifier(v) {
    if (arguments.length === 1) {
      this.fields.teamIdentifier = v;
      return this;
    }
    return this.fields.teamIdentifier;
  }

  associatedStoreIdentifiers(v) {
    if (arguments.length === 1) {
      this.fields.associatedStoreIdentifiers = v;
      return this;
    }
    return this.fields.associatedStoreIdentifiers;
  }

  description(v) {
    if (arguments.length === 1) {
      this.fields.description = v;
      return this;
    }
    return this.fields.description;
  }

  backgroundColor(v) {
    if (arguments.length === 1) {
      this.fields.backgroundColor = Template.convertToRgb(v);
      return this;
    }
    return this.fields.backgroundColor;
  }

  foregroundColor(v) {
    if (arguments.length === 1) {
      this.fields.foregroundColor = Template.convertToRgb(v);
      return this;
    }
    return this.fields.foregroundColor;
  }

  labelColor(v) {
    if (arguments.length === 1) {
      this.fields.labelColor = Template.convertToRgb(v);
      return this;
    }
    return this.fields.labelColor;
  }

  logoText(v) {
    if (arguments.length === 1) {
      this.fields.logoText = v;
      return this;
    }
    return this.fields.logoText;
  }

  organizationName(v) {
    if (arguments.length === 1) {
      this.fields.organizationName = v;
      return this;
    }
    return this.fields.organizationName;
  }

  groupingIdentifier(v) {
    if (arguments.length === 1) {
      this.fields.groupingIdentifier = v;
      return this;
    }
    return this.fields.groupingIdentifier;
  }

  /**
   * sets or gets suppressStripShine
   *
   * @param {boolean?} v
   * @returns {Template | boolean}
   * @memberof Template
   */
  suppressStripShine(v) {
    if (arguments.length === 1) {
      if (typeof v !== 'boolean')
        throw new Error('suppressStripShine value must be a boolean!');
      this.fields.suppressStripShine = v;
      return this;
    }
    return this.fields.suppressStripShine;
  }

  /**
   * gets or sets webServiceURL
   *
   * @param {URL | string} v
   * @returns {Template | string}
   * @memberof Template
   */
  webServiceURL(v) {
    if (arguments.length === 1) {
      // validating URL, it will throw on bad value
      const url = v instanceof URL ? v : new URL(v);
      if (url.protocol !== 'https:')
        throw new Error(`webServiceURL must be on HTTPS!`);
      this.fields.webServiceURL = url.toString();
      return this;
    }
    return this.fields.webServiceURL;
  }

  authenticationToken(v) {
    if (arguments.length === 1) {
      if (typeof v !== 'string' || v.length < 16)
        throw new Error(
          `authenticationToken must be a string and have more than 16 characters`,
        );
      this.fields.authenticationToken = v;
      return this;
    }
    return this.fields.authenticationToken;
  }

  /**
   * Sets path to directory containing keys and password for accessing keys.
   *
   * @param {string} keysPath - Path to directory containing key files (default is 'keys')
   * @param {string} password - Password to use with keys
   * @memberof Template
   */
  keys(keysPath, password) {
    if (typeof keysPath === 'string') this.keysPath = keysPath;
    if (password) this.password = password;
  }

  /**
   * Create a new pass from a template.
   *
   * @param {Object} fields
   * @returns {Pass}
   * @memberof Template
   */
  createPass(fields = {}) {
    // Combine template and pass fields
    return new Pass(this, Object.assign({}, this.fields, fields), this.images);
  }

  /**
   * Loads Template, images and key from a given path
   *
   * @static
   * @param {string} folderPath
   * @param {string} keyPassword - optional key password
   * @returns {Promise.<Template>}
   * @throws - if given folder doesn't contain pass.json or it is in invalid format
   * @memberof Template
   */
  static async load(folderPath, keyPassword) {
    // Check if the path is accessible directory actually
    const stats = await stat(folderPath);
    if (!stats.isDirectory())
      throw new Error(`Path ${folderPath} must be a directory!`);

    // getting main JSON file
    const passJson = JSON.parse(
      await readFile(join(folderPath, 'pass.json'), 'utf8'),
    );

    // Trying to detect the type of pass
    let type;
    if (
      !PASS_STYLES.some(t => {
        if (t in passJson) {
          type = t;
          return true;
        }
        return false;
      })
    )
      throw new Error('Unknown pass style!');

    const template = new Template(type, passJson);

    // load images from the same folder
    await template.images.loadFromDirectory(folderPath);

    // checking if there is a key - must be named ${passTypeIdentifier}.pem
    const typeIdentifier = passJson.passTypeIdentifier;
    const keyName = `${typeIdentifier.replace(/^pass\./, '')}.pem`;
    try {
      const keyStat = await stat(join(folderPath, keyName));
      if (keyStat.isFile()) template.keys(folderPath, keyPassword);
      // eslint-disable-next-line no-empty
    } catch (_) {}
    // done
    return template;
  }
}

module.exports = Template;
