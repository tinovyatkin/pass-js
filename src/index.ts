import * as constants from './constants';

export { constants };
export { Template } from './template';
export { Pass } from './pass';
