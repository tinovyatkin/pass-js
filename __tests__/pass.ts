'use strict';

import { createHash } from 'crypto';
import { unlinkSync, mkdtempSync, writeFileSync } from 'fs';
import { tmpdir } from 'os';
import * as path from 'path';
import { execFileSync } from 'child_process';

import * as constants from '../src/constants';
import { Template } from '../src/template';

// Clone all the fields in object, except the named field, and return a new
// object.
//
// object - Object to clone
// field  - Except this field
function cloneExcept(object, field) {
  const clone = {};
  for (const key in object) {
    if (key !== field) clone[key] = object[key];
  }
  return clone;
}

function unzip(zipFile, filename): Buffer {
  return execFileSync('unzip', ['-p', zipFile, filename], {
    encoding: 'buffer',
  });
}

const template = new Template('coupon', {
  passTypeIdentifier: 'pass.com.example.passbook',
  teamIdentifier: 'MXL',
  labelColor: 'red',
});

const fields = {
  serialNumber: '123456',
  organizationName: 'Acme flowers',
  description: '20% of black roses',
};

describe('Pass', () => {
  beforeAll(async () => {
    template.setCertificate(process.env.APPLE_PASS_CERTIFICATE);
    template.setPrivateKey(
      process.env.APPLE_PASS_PRIVATE_KEY,
      process.env.APPLE_PASS_KEY_PASSWORD,
    );
  });
  it('from template', () => {
    const pass = template.createPass();

    // should copy template fields
    expect(pass.passTypeIdentifier).toBe('pass.com.example.passbook');

    // should start with no images
    expect(pass.images.count).toBe(0);

    // should create a structure based on style
    expect(pass.style).toBe('coupon');
    expect(pass.labelColor).toEqual([255, 0, 0]);
  });

  //
  it('barcodes as Array', () => {
    const pass = template.createPass(cloneExcept(fields, 'serialNumber'));
    expect(() => {
      pass.barcodes = [
        {
          format: 'PKBarcodeFormatQR',
          message: 'Barcode message',
          messageEncoding: 'iso-8859-1',
        },
      ];
    }).not.toThrow();
    expect(() => {
      pass.barcodes = 'byaka';
    }).toThrow();
  });

  it('without serial number should not be valid', () => {
    const pass = template.createPass(cloneExcept(fields, 'serialNumber'));
    expect(() => pass.validate()).toThrow('serialNumber is required in a Pass');
  });

  it('without organization name should not be valid', () => {
    const pass = template.createPass(cloneExcept(fields, 'organizationName'));
    expect(() => pass.validate()).toThrow(
      'organizationName is required in a Pass',
    );
  });

  it('without description should not be valid', () => {
    const pass = template.createPass(cloneExcept(fields, 'description'));
    expect(() => pass.validate()).toThrow('description is required in a Pass');
  });

  it('without icon.png should not be valid', () => {
    const pass = template.createPass(fields);
    expect(() => pass.validate()).toThrow('Missing required image icon.png');
  });

  it('without logo.png should not be valid', async () => {
    const pass = template.createPass(fields);
    await pass.images.setImage(
      'icon',
      path.resolve(__dirname, './resources/icon.png'),
      undefined,
      'en-US',
    );

    expect.assertions(1);
    try {
      await pass.asBuffer();
    } catch (err) {
      expect(err).toHaveProperty('message', 'Missing required image logo.png');
    }
  });

  it('boarding pass has string-only property in structure fields', async () => {
    const templ = await Template.load(
      path.resolve(__dirname, './resources/passes/BoardingPass.pass/'),
    );
    expect(templ.style).toBe('boardingPass');
    // ensures it parsed color read from JSON
    expect(templ.backgroundColor).toEqual([50, 91, 185]);
    // ensure it parses well fields
    expect(templ.backFields.size).toBe(2);
    expect(templ.auxiliaryFields.size).toBe(4);
    expect(templ.relevantDate).toBeDate();
    expect(templ.relevantDate.getFullYear()).toBe(2012);
    expect(templ.barcodes).toBeArrayOfObjects();
    // switching transit type
    const pass = templ.createPass();
    expect(pass.transitType).toBe(constants.TRANSIT.AIR);
    pass.transitType = constants.TRANSIT.BUS;
    expect(pass.transitType).toBe(constants.TRANSIT.BUS);
    expect(pass.toJSON()).toMatchObject({
      boardingPass: expect.objectContaining({
        transitType: constants.TRANSIT.BUS,
      }),
    });
  });

  it('should convert back to the same pass.json', async () => {
    const t = await Template.load(
      path.resolve(__dirname, './resources/passes/Event.pass'),
    );
    expect(require('./resources/passes/Event.pass/pass.json')).toMatchObject(
      JSON.parse(JSON.stringify(t)),
    );
  });

  it('asBuffer returns buffer with ZIP file', async () => {
    const pass = template.createPass(fields);
    await pass.images.loadFromDirectory(path.resolve(__dirname, './resources'));
    pass.headerFields.add({ key: 'date', value: 'Date', label: 'Nov 1' });
    pass.primaryFields.add({
      key: 'location',
      label: 'Place',
      value: 'High ground',
    });
    const tmd = mkdtempSync(`${tmpdir()}${path.sep}`);
    const passFileName = path.join(tmd, 'pass.pkpass');
    const buf = await pass.asBuffer();
    expect(Buffer.isBuffer(buf)).toBeTruthy();
    writeFileSync(passFileName, buf);
    // test that result is valid ZIP at least
    const stdout = execFileSync('unzip', ['-t', passFileName], {
      encoding: 'utf8',
    });
    unlinkSync(passFileName);
    expect(stdout).toContain('No errors detected in compressed data');
  });
});

describe('generated', () => {
  const tmd = mkdtempSync(`${tmpdir()}${path.sep}`);
  const passFileName = path.join(tmd, 'pass.pkpass');

  beforeAll(async () => {
    jest.setTimeout(100000);
    const pass = template.createPass(fields);
    await pass.images.loadFromDirectory(path.resolve(__dirname, './resources'));
    pass.headerFields.add({ key: 'date', label: 'Date', value: 'Nov 1' });
    pass.primaryFields.add({
      key: 'location',
      label: 'Place',
      value: 'High ground',
    });
    writeFileSync(passFileName, await pass.asBuffer());
  });

  afterAll(async () => {
    unlinkSync(passFileName);
  });

  it('should be a valid ZIP', async () => {
    const stdout = execFileSync('unzip', ['-t', passFileName], {
      encoding: 'utf8',
    });
    expect(stdout).toContain('No errors detected in compressed data');
  });

  it('should contain pass.json', async () => {
    const res = JSON.parse(unzip(passFileName, 'pass.json').toString('utf8'));
    expect(res).toMatchSnapshot();
  });

  it('should contain a manifest', async () => {
    const res = JSON.parse(
      unzip(passFileName, 'manifest.json').toString('utf8'),
    );
    expect(res).toMatchSnapshot();
  });

  // this test depends on MacOS specific signpass, so, run only on MacOS
  if (process.platform === 'darwin') {
    it('should contain a signature', async () => {
      const stdout = execFileSync(
        path.resolve(__dirname, './resources/bin/signpass'),
        ['-v', passFileName],
        { encoding: 'utf8' },
      );
      expect(stdout).toContain('*** SUCCEEDED ***');
    });
  }

  it('should contain the icon', async () => {
    const buffer = unzip(passFileName, 'icon.png');
    expect(
      createHash('sha1')
        .update(buffer)
        .digest('hex'),
    ).toBe('e0f0bcd503f6117bce6a1a3ff8a68e36d26ae47f');
  });

  it('should contain the logo', async () => {
    const buffer = unzip(passFileName, 'logo.png');
    expect(
      createHash('sha1')
        .update(buffer)
        .digest('hex'),
    ).toBe('abc97e3b2bc3b0e412ca4a853ba5fd90fe063551');
  });
});
