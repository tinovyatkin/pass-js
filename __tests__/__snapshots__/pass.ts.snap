// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generated should contain a manifest 1`] = `
Object {
  "en-US.lproj/icon.png": "e0f0bcd503f6117bce6a1a3ff8a68e36d26ae47f",
  "icon.png": "e0f0bcd503f6117bce6a1a3ff8a68e36d26ae47f",
  "icon@2x.png": "10e4a72dbb02cc526cef967420553b459ccf2b9e",
  "logo.png": "abc97e3b2bc3b0e412ca4a853ba5fd90fe063551",
  "logo@2x.png": "87ca39ddc347646b5625062a349de4d3f06714ac",
  "pass.json": "7f292a16db3f608e6e3e5017eb2bcafd8e035fbb",
  "strip.png": "68fc532d6c76e7c6c0dbb9b45165e62fbb8e9e32",
  "strip@2x.png": "17e4f5598362d21f92aa75bc66e2011a2310f48e",
  "thumbnail.png": "e199fc0e2839ad5698b206d5f4b7d8cb2418927c",
  "thumbnail@2x.png": "ac640c623741c0081fb1592d6353ebb03122244f",
}
`;

exports[`generated should contain pass.json 1`] = `
Object {
  "coupon": Object {
    "headerFields": Array [
      Object {
        "key": "date",
        "label": "Date",
        "value": "Nov 1",
      },
    ],
    "primaryFields": Array [
      Object {
        "key": "location",
        "label": "Place",
        "value": "High ground",
      },
    ],
  },
  "description": "20% of black roses",
  "formatVersion": 1,
  "labelColor": "rgb(255, 0, 0)",
  "organizationName": "Acme flowers",
  "passTypeIdentifier": "pass.com.example.passbook",
  "serialNumber": "123456",
  "teamIdentifier": "MXL",
}
`;
